src/company_gui.cpp

```diff
@@ -2066,6 +2088,23 @@ struct CompanyInfrastructureWindow : Window
 		}
 	}
 
+	void OnClick([[maybe_unused]] Point pt, WidgetID widget, [[maybe_unused]] int click_count) override
+	{
+		if (widget != WID_CI_HIGHLIGHT_INFRASTRUCTURE) return;
+
+		/* Toggle between highlight and clear */
+		CompanyID window_company_id = (CompanyID)this->window_number;
+		if(_viewport_company_to_highlight_infrastructure != window_company_id) {
+			/* highlight tiles of this company */
+			_viewport_company_to_highlight_infrastructure = window_company_id;
+			MarkWholeScreenDirty();
```

Comment:
This `MarkWholeScreenDirty();` could be outside the if/else block since it's called for both branches.